relay-server/src/processing/transactions/mod.rs

```diff
@@ -0,0 +1,592 @@
+        }
+
+        // Check attachment limits:
+        for (category, quantity) in span_quantities {
+            let limits = rate_limiter
+                .try_consume(scoping.item(category), quantity)
+                .await;
+
+            if !limits.is_empty() {
+                self.modify(|this, record_keeper| {
+                    record_keeper.reject_err(
+                        Error::from(limits),
+                        std::mem::take(&mut this.extracted_spans),
+                    );
+                });
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Default)]
+struct Flags {
+    metrics_extracted: bool,
+    spans_extracted: bool,
+    fully_normalized: bool,
+}
+
+#[derive(Debug)]
+struct Transaction(Annotated<Event>);
+
+impl Counted for Transaction {
+    fn quantities(&self) -> Quantities {
+        match self.0.value() {
+            Some(_) => smallvec![
+                (DataCategory::TransactionIndexed, 1),
+                (DataCategory::Transaction, 1),
+            ],
+            None => smallvec![],
+        }
+    }
+}
+
+// pub struct CountSpans<'a>(&'a Event);
+
+// impl Counted for CountSpans<'_> {
+//     fn quantities(&self) -> Quantities {
+//         let mut quantities = self.0.quantities();
+//         quantities.extend(self.0.spans.quantities());
+//         quantities
+//     }
+// }
+
+#[derive(Debug)]
+pub struct TransactionOutput(Managed<ExpandedTransaction>);
+
+impl Forward for TransactionOutput {
+    fn serialize_envelope(
+        self,
+        ctx: super::ForwardContext<'_>,
+    ) -> Result<Managed<Box<Envelope>>, Rejected<()>> {
+        self.0.try_map(|output, _| {
+            output
+                .serialize_envelope()
+                .map_err(drop)
+                .with_outcome(Outcome::Invalid(DiscardReason::Internal))
+        })
+    }
+
+    #[cfg(feature = "processing")]
+    fn forward_store(
+        self,
+        s: &relay_system::Addr<crate::services::store::Store>,
+        ctx: super::ForwardContext<'_>,
+    ) -> Result<(), Rejected<()>> {
+        let Self(output) = self;
+        let outcome_addr = output.outcome_addr().clone();
+
+        // pass ownership to store service.
+        output
+            .try_accept(|output| {
+                // TODO: split out spans here.
+                let envelope = output
+                    .serialize_envelope()
+                    .map_err(|_| Outcome::Invalid(DiscardReason::Internal))?;
+
+                let envelope = ManagedEnvelope::new(envelope, outcome_addr);
+                let envelope = TypedEnvelope::<ProcessingGroup>::try_from((
+                    envelope,
+                    ProcessingGroup::Transaction,
+                ))
+                .map_err(|_| Outcome::Invalid(DiscardReason::Internal))?
+                .into_processed();
+
+                s.send(StoreEnvelope { envelope });
+                Ok::<_, Outcome>(())
+            })
+            .map_err(|e| e.map(|_| ()))
```

Comment:
```suggestion
        let envelope: ManagedEnvelope = output
            .try_map(|tx, _| {
                tx.serialize_envelope()
                    .map_err(|_| (Outcome::Invalid(DiscardReason::Internal), ()))
            })?
            .into();

        s.send(StoreEnvelope {
            envelope: envelope.into_processed(),
        });

        Ok(())
```

This also means `outcome_addr` is no longer needed.